### PR TITLE
feat(FN-3196): persist form values on the edit a payment screen

### PR DESCRIPTION
**Pull request by sw-mattw**
## Introduction :pencil2:
This PR introduces functionality to persist user input on the edit payment screen. It preserves form values and checkbox selections when users interact with another form on the same page, ensuring that no previously entered information is accidentally discarded.

## Resolution :heavy_check_mark:
- Added persistence of form values and checkbox selections when removing fees from payments.

## Miscellaneous :heavy_plus_sign:
- Modified E2E tests to check for value persistence.

### CODE REVIEW

File: trade-finance-manager-ui/server/controllers/utilisation-reports/edit-payment/index.ts [hunk 1]
```diff
@@ -52,8 +63,19 @@ export const postEditPayment = async (req: PostEditPaymentRequest, res: Response
       return res.redirect(`/utilisation-reports/${reportId}`);
     }
 
+    const checkedCheckboxIds = getEditPaymentsCheckboxIdsFromObjectKeys(req.body);
+    const checkedCheckboxIdsRecord = mapCheckedCheckboxesToRecord(checkedCheckboxIds);
+    const isCheckboxChecked = (checkboxId: string): boolean => Boolean(checkedCheckboxIdsRecord[checkboxId]);
```
Comment (RLavender98) [inline trade-finance-manager-ui/server/controllers/utilisation-reports/edit-payment/index.ts:68]: Could we just create a set of checkedCheckboxIds and then check for inclusion in the set? I think that would be simpler and then we wouldn't need to export and make common the `mappedCheckedCheckboxesToRecord` method?

File: trade-finance-manager-ui/server/controllers/utilisation-reports/remove-fees-from-payment/index.test.ts [hunk 2]
```diff
@@ -23,20 +24,49 @@ describe('controllers/utilisation-reports/remove-fees-from-payment', () => {
       jest.resetAllMocks();
     });
 
-    it('removes the selected fee records from the payment', async () => {
+    it('sets the request session editPaymentFormValues field', async () => {
```
Comment (RLavender98) [inline trade-finance-manager-ui/server/controllers/utilisation-reports/remove-fees-from-payment/index.test.ts:27]: Minor comment but I think this should be included in the redirect test e.g. `redirects to '/utilisation-reports/:reportId/edit-payment/:paymentId' with the edit payment form values when removing fee records is successful`

**Review by RLavender98 — COMMENTED** [review]
Mostly just a few comments around names/types!

File: trade-finance-manager-ui/server/controllers/utilisation-reports/edit-payment/get-and-clear-fields-from-redirect-session-data.test.ts [hunk 3]
```diff
@@ -1,53 +1,77 @@
...
       },
     });
 
   const assertSessionHasBeenCleared = (req: ReturnType<typeof getMockRequest>) => {
     expect(req.session.removeFeesFromPaymentErrorKey).toBeUndefined();
+    expect(req.session.editPaymentFormValues).toBeUndefined();
   };
 
-  it('clears the session and returns an undefined errorSummary when the session error keys are undefined', () => {
+  it('clears the session and returns provided form values and an undefined errorSummary when the session error keys are undefined', () => {
```
[the remark this answers is not in the document]
Comment (sw-mattw) [reply trade-finance-manager-ui/server/controllers/utilisation-reports/edit-payment/get-and-clear-fields-from-redirect-session-data.test.ts:42]: Ah yes, added thanks

File: trade-finance-manager-ui/server/controllers/utilisation-reports/edit-payment/index.test.ts [hunk 4]
```diff
@@ -84,10 +84,89 @@ describe('controllers/utilisation-reports/edit-payment', () => {
...
+        await getEditPayment(req, res);
+
+        // Assert
+        const viewModel = res._getRenderData() as EditPaymentViewModel;
+        expect(viewModel.feeRecords).toHaveLength(feeRecords.length);
+        viewModel.feeRecords.forEach(({ isChecked }) => expect(isChecked).toBe(true));
+      });
+    });
+
+    it('sets the render view model formValues based on passed in session data', async () => {
```
[the remark this answers is not in the document]
Comment (sw-mattw) [reply trade-finance-manager-ui/server/controllers/utilisation-reports/edit-payment/index.test.ts:157]: Removed references from individual tests as session data is explicitly mentioned in the surrounding describe blocks anyway

File: trade-finance-manager-ui/server/controllers/utilisation-reports/helpers/edit-payment-helper.test.ts [hunk 5]
```diff
@@ -218,28 +219,36 @@ describe('edit-payment-helper', () => {
...
+    it('sets the render view model feeRecords isChecked value using the provided isCheckboxChecked function for every fee record', () => {
       // Arrange
-      const feeRecords = [aFeeRecord(), aFeeRecord(), aFeeRecord(), aFeeRecord()];
+      const feeRecords = [
+        { ...aFeeRecord(), id: 1 },
+        { ...aFeeRecord(), id: 2 },
+      ];
+      const checkedFeeRecordIds = ['feeRecordId-1'];
+
+      const mockIsCheckboxChecked = jest.fn().mockImplementation((id: string) => checkedFeeRecordIds.includes(id));
```
[the remark this answers is not in the document]
Comment (sw-mattw) [reply trade-finance-manager-ui/server/controllers/utilisation-reports/helpers/edit-payment-helper.test.ts]: Changed to this use jest-when

File: trade-finance-manager-ui/server/middleware/validatePostRemoveFeesFromPaymentRequestBody/index.ts [hunk 6]
```diff
@@ -1,22 +1,31 @@
...
 };
 
-const renderProblemWithServiceView = (req: Request, res: Response) => {
+const renderProblemWithServiceView = (req: RemoveFeesFromPaymentRequest, res: Response) => {
   const { user } = asUserSession(req.session);
   return res.render('_partials/problem-with-service.njk', { user });
 };
 
-export const validatePostRemoveFeesFromPaymentRequestBody = (req: Request, res: Response, next: NextFunction) => {
+export const validatePostRemoveFeesFromPaymentRequestBody = (req: RemoveFeesFromPaymentRequest, res: Response, next: NextFunction) => {
```
[the remark this answers is not in the document]
Comment (sw-mattw) [reply trade-finance-manager-ui/server/middleware/validatePostRemoveFeesFromPaymentRequestBody/index.ts]: Changed the typing on this now and added a unit test for the case of no edit form values in the request body

File: trade-finance-manager-ui/server/controllers/utilisation-reports/edit-payment/index.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (RLavender98) [reply trade-finance-manager-ui/server/controllers/utilisation-reports/edit-payment/index.test.ts:157]: I'm pretty sure this test is not in that describe block and so is at the same level as the ones below asserting on setting the forma values using the api values

[the remark this answers is not in the document]
Comment (sw-mattw) [reply trade-finance-manager-ui/server/controllers/utilisation-reports/edit-payment/index.test.ts]: Ah yep sorry, grouped these cases in 1433ad46267c4d2e598938df683a95c82d60b141

**Review by RLavender98 — APPROVED** [review]

**Pull request merged**